database/api.go

```diff
@@ -176,5 +193,23 @@ func (d *dbapi) runInTransaction(fn func(*sqlx.Tx) error) error {
 		return err
 	}
 
-	return d.CommitOrRollback(fn(d.tx))
+	return d.CommitOrRollback(d.slowlog(func() error {
+		return fn(d.tx)
+	}))
+}
+
+func (db *dbapi) slowlog(fn func() error) error {
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		err := errors.New("query execution time: %s transaction: %s", elapsed, db.txID)
+
+		if elapsed > db.slowQueryThreshold {
+			_ = db.log.Error(err)
+		} else {
+			_ = db.log.Debug(err)
+		}
+	}()
+
+	return fn()
```

Comment:
we should not be using a defer here since we could just call the function and then do the logic, there is only a single exit point. Also, we don't get anything by having this logic in a separate function, so we should just put it in runInTransaction and have this function just do the logging, e.g. 
``` golang
func (d *dbapi) runInTransaction(fn func(*sqlx.Tx) error) error {
	var (
		start  time.Time
		err    error
		commit bool
	)
	if d.tx == nil {
		commit = true
		err = d.Begin()
	}
	if nil != err {
		return err
	}

	start = time.Now()
	err = fn(d.tx)
	d.logExecutionTime(time.Since(start))

	if commit {
		err = d.CommitOrRollback(err)
	}
	return err
}

func (db *dbapi) logExecutionTime(elapsed time.Duration) {
	err := errors.New("query execution time: %s transaction: %s", elapsed, db.txID)
	if elapsed > db.slowQueryThreshold {
		_ = db.log.Error(err)
	} else {
		_ = db.log.Debug(err)
	}
}
```

Reply:
The stack trace is not going to be useful for us. What we can do is wrap the transaction in an interface with QueryRowx, Select, NamedExec, Exec, Commit, and Rollback and pass that into the CrudTx functions. Since those functions are passed the actual query and our queries all parameterized we can just log them without worrying about exposing sensitive information. Then we can log any slow queries inside the transaction wrapper.
Something like:
```
type slowQueryLoggerTx struct {
    log 
    slowQueryDuration time.Duration
    tx *sqlx.Tx
}
func (sqlt slowQueryLoggerTx) Exec(...) (...) {
    start = time.Now()
    r, err = tx.Exec(args)
    sqlt.LogSlowQuery(time.Since(start))
    return r, err
}

func (d *dpapi) wrap(tx *sqlx.Tx) Transaction {
    return &slowQueryLoggerTx{...}
}

func (d *dbapi) runInTransaction(fn func(*sqlx.Tx) error) error {
    ...
    err = fn(d.Wrap(tx))
    ...
}
```